torch/testing/_internal/common_utils.py

```diff
@@ -1072,24 +1072,112 @@ def setUp(self):
+            counts[1:n_rows - n + 1].fill_(m)
+
+        if N:
+            q, r = divmod(nnz - n * n_cols - m * (n_rows - n),
+                          (n_cols - m) * (n_cols - m + 1) // 2)
+            p = 1 + q * (n_cols - m + 1)
+            if sys.version_info >= (3, 8):
+                k = math.isqrt(2 * r)
+            else:
+                # math.isqrt(x) is available starting from Python 3.8.
+                # Here we use int(math.sqrt(x)) as an approximation
+                # that appers to give exaxt result for all x values
+                # less than 2**35, at least, the upper limit of x is
+                # TBD.
+                k = int(math.sqrt(2 * r))
+            if k * (k + 1) > 2 * r:
+                k -= 1
+            corr = r - k * (k + 1) // 2
+            assert not ((p > 1) and (m > 0))  # full sawteeth are never on top of a bottom rectangle
+            # sequence of full sawteeth:
+            counts[1:p] = torch.arange(p - 1, dtype=dtype, device=counts.device) % (n_cols - m + 1)
+            # incomplete sawtooth:
+            counts[p:p + k + 1] += torch.arange(k + 1, dtype=dtype, device=counts.device)
+            # correction:
+            counts[p] += corr
+        else:
+            # correction:
+            counts[1] = nnz - n * n_cols - m * (n_rows - n)
+
+        if random:
+            perm = torch.randperm(n_rows, device=counts.device)
+            counts[1:] = counts[1:][perm]
+
+        crow_indices = counts
+        crow_indices.cumsum_(dim=0)
+        return crow_indices.to(device=device)
+
     def genSparseCSRTensor(self, size, nnz, *, device, dtype, index_dtype):
         sparse_dim = 2
         assert all(size[d] > 0 for d in range(sparse_dim)) or nnz == 0, 'invalid arguments'
         assert len(size) == sparse_dim
 
         def random_sparse_csr(n_rows, n_cols, nnz):
-            nnz_per_row = nnz // n_rows
-            if nnz_per_row > 0:
-                crow_indices = torch.zeros(n_rows + 1, dtype=index_dtype)
-                crow_indices[1:] = nnz_per_row
-                crow_indices.cumsum_(dim=0)
-                col_indices = torch.randint(0, n_cols, size=[nnz_per_row * n_rows], dtype=index_dtype, device=device)
-            else:
-                crow_indices = torch.zeros(n_rows + 1, dtype=index_dtype)
-                crow_indices[1:nnz + 1] = 1
-                crow_indices.cumsum_(dim=0)
-                col_indices = torch.randint(0, n_cols, size=[nnz], dtype=index_dtype, device=device)
-            nnz = col_indices.shape[0]
+            crow_indices = self._make_crow_indices(n_rows, n_cols, nnz, device=device, dtype=index_dtype)
+            col_indices = torch.zeros(nnz, dtype=index_dtype, device=device)
+            n = 0
+            for i in range(n_rows):
+                count = crow_indices[i + 1] - crow_indices[i]
+                col_indices[n:n + count], _ = torch.sort(torch.randperm(n_cols, dtype=index_dtype, device=device)[:count])
+                n += count
```

Comment:
Sadly PyTorch doesn't have specialized sampling of integers without replacement and we need to do permutation followed by indexing.

Reply:
Assuming the size of ncols is relatively small [random.choices](https://docs.python.org/3/library/random.html#random.choices) might just be enough (and probably faster than randperm + indexing).

Reply:
We cannot use `random.choices` because it generates repeated values.

Here's a quick benchmark for three other possible choices:
```python
In [18]: population = list(range(100))

In [19]: %timeit random.shuffle(population); torch.tensor(population[:10])
29.1 µs ± 3.45 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [20]: %timeit torch.tensor(random.sample(population, k=10), dtype=int)
5.63 µs ± 103 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [21]: %timeit torch.sort(torch.randperm(100, dtype=int)[:10])[0]
7.22 µs ± 28.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
Using population size 10000 (that corresponds to `n_cols`) gave about the same timings for the last two. However, for increasing crow_indices differences (that corresponds to increasing nnz), the last approach gives the best performance:
```
In [32]: population = list(range(10000))

In [33]: %timeit torch.tensor(random.sample(population, k=100), dtype=int)
33.8 µs ± 99.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [34]: %timeit torch.sort(torch.randperm(100, dtype=int)[:100])[0]
9.72 µs ± 22.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

So, `randomperm + sort + indexing` is not that bad after all and I think we can stick to it because it (i) scales better and (ii) uses torch-only collections. Notice that creating a torch tensor from a 100-list takes about 2.5 µs, so the performance timing 9.72 µs is quite reasonable.